### DISCUSSION TITLE
Configure DBTCliResource to use specific target within profile at run time

### DISCUSSION THREAD

**Discussion by MeganBeckett** [dagster-io/dagster#22426] [Q&A] [answered]
Hi there, I was previously able to configure a dbt_cli_resource so I can specify the target within the profile to use. The reason for this is the different targets specify different schemas (ie. dev, prod, etc).
I had something like this:

```
@configured(
        dbt_cli_resource, 
        config_schema={
            "profile": Field(str, description="Which profile to load in profiles.yml. Overrides setting in dbt_project.yml. ie. which database."),
            "target": Field(str, description="Which target to load for the given profile, ie. which schema.")
            }
        )
def custom_dbt_cli_resource(config):
    return {
        "profile": config["profile"],
        "target": config["target"],
        "project-dir": DBT_PROJECT_DIR,
        "profiles-dir": DBT_PROFILES_DIR
    }

dbt_assets = with_resources(
    load_assets_from_dbt_project(
        project_dir=DBT_PROJECT_DIR,
        node_info_to_group_fn=lambda _: "dbt_assets",
        use_build_command=True),
    {
        "dbt": custom_dbt_cli_resource
    },
)
```

However, with the new DBTClieResource, it accepts a manifest.json and this file is generated with dbt parse where you have to specify a target up front so then the schema in manifest.json is already defined as dev or prod.
How can I make this dynamic again so that I can specify the target (ie. then which schema to run the pipeline for) at run time and in the UI?

**Comment by rexledesma** [comment, marked as answer]
You can use Dagster config, and then pass in the profile and target in the CLI execution arguments.

See an example with how we use this to pass in `--full-refresh` at run time: https://docs.dagster.io/integrations/dbt/reference#using-config-with-dbt_assets

**Reply by MeganBeckett** [reply]
Hi again, I implemented the following:

```
@dbt_assets(manifest=dbt_manifest_path)
def dr_dbt_assets(
    context: AssetExecutionContext, dbt: DbtCliResource, config: DbtConfig
    ):
    # Create the CLI build command by adding in custom tags
    dbt_build_args = ["build"]

    # Do a full refresh each time of the views
    if config.full_refresh:
        dbt_build_args += ["--full-refresh"]

    # Specify the target in the profile to use, either dev or prod
    dbt_build_args += [f"--target {config.target}"]

    yield from dbt.cli(dbt_build_args, context=context).stream()
```

However, when running the pipeline, I am now getting an error about the `--select` argument and that it is not set and so the default is used, but this fails:

![image](https://github.com/dagster-io/dagster/assets/3185323/5600c723-b866-402e-9361-f50449ca972f)

And the error:

![image](https://github.com/dagster-io/dagster/assets/3185323/5b6ee1a7-1da0-4ec0-be8f-8f36542c4fe5)

Do I now also have to specify a `--select` argument for the CLI command and/or why does the default not work?

**Reply by MeganBeckett** [reply]
Following on from this, I tried trhe following by specifying a tag that I have in the dbt models for `tag:daily`, but it still does not work and I get the same error, even with the `--select` argument I provided in the CLI command:

Updated code:

```
@dbt_assets(manifest=dbt_manifest_path)
def dr_dbt_assets(
    context: AssetExecutionContext, dbt: DbtCliResource, config: DbtConfig
    ):
    # Create the CLI build command by adding in custom tags
    dbt_build_args = ["build"]

    # Do a full refresh each time of the views
    if config.full_refresh:
        dbt_build_args += ["--full-refresh"]

    # Specify the target in the profile to use, either dev or prod
    dbt_build_args += [f"--target {config.target} --select tag:daily"]

    yield from dbt.cli(dbt_build_args, context=context).stream()
```

Info message again:
![image](https://github.com/dagster-io/dagster/assets/3185323/d2eec901-5228-4fb6-af00-870f806251e5)

Same error: 
![image](https://github.com/dagster-io/dagster/assets/3185323/8c151a33-dad1-4a74-acd1-f02dc4be7466)

**Reply by rexledesma** [reply]
You need to split up the arguments. We execute `dbt` using [`subprocess`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen), so you need to correctly tokenized your arguments.

Could you try the following?

```diff
@dbt_assets(manifest=dbt_manifest_path)
def dr_dbt_assets(
    context: AssetExecutionContext, dbt: DbtCliResource, config: DbtConfig
    ):
    # Create the CLI build command by adding in custom tags
    dbt_build_args = ["build"]

    # Do a full refresh each time of the views
    if config.full_refresh:
        dbt_build_args += ["--full-refresh"]

    # Specify the target in the profile to use, either dev or prod
-   dbt_build_args += [f"--target {config.target}"]
+   dbt_build_args += ["--target", config.target]

    yield from dbt.cli(dbt_build_args, context=context).stream()
```

**Reply by MeganBeckett** [reply]
Ah ok, that works now, thanks.

Just a small modification to move the f string for reference for anyone else, so:

```
dbt_build_args += ["--target", f"{config.target}"]
```
